### PR TITLE
Add interior surface sections to Envelope Summary Report

### LINKED ISSUE #8007: Envelope Summary report should include additional Interior sections
Issue overview
--------------
The Envelope Summary output report currently includes the following sections:
- Opaque Exterior
- Exterior Fenestration
- Interior Fenestration
- Exterior Door

In residential buildings, it's common to have interior opaque surfaces (e.g., attic floors, basement ceilings, walls between conditioned space and garage) through which the heat transfer is just as important as exterior opaque surfaces. It is currently not possible to view properties for these surfaces in the output.

To address this, I suggest adding two additional sections:
- Opaque Interior
- Interior Door

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by yzhou601**
Pull request overview
---------------------
 - Fixes #8007 
 - Added interior sections to envelope summery report

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [x] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 1]
```diff
@@ -1169,9 +1169,37 @@ namespace HeatBalanceSurfaceManager {
...
+                        if ((curAzimuth >= 315.0) || (curAzimuth < 45.0)) {
+                            PreDefTableEntry(pdchIntOpDir, surfName, "N");
+                        } else if ((curAzimuth >= 45.0) && (curAzimuth < 135.0)) {
+                            PreDefTableEntry(pdchIntOpDir, surfName, "E");
+                        } else if ((curAzimuth >= 135.0) && (curAzimuth < 225.0)) {
+                            PreDefTableEntry(pdchIntOpDir, surfName, "S");
+                        } else if ((curAzimuth >= 225.0) && (curAzimuth < 315.0)) {
+                            PreDefTableEntry(pdchIntOpDir, surfName, "W");
+                        }
+                    }
```
Comment (Nigusse) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:1175]: *Opaque Interior* surfaces data are retrieved here and passed to PreDefTableEntry() function to create the tables. Applies zone or zoneList multiplier if available. Cardinal direction are not set for interior floors and ceilings. This looks good.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 2]
```diff
@@ -1210,6 +1239,15 @@ namespace HeatBalanceSurfaceManager {
...
                     }
+                }else if (Surface(iSurf).Class == SurfaceClass_Door) {
+                    surfName = Surface(iSurf).Name;
+                    curCons = Surface(iSurf).Construction;
+                    PreDefTableEntry(pdchIntDrCons, surfName, dataConstruction.Construct(curCons).Name);
+                    PreDefTableEntry(pdchIntDrUfactNoFilm, surfName, NominalU(Surface(iSurf).Construction), 3);
+                    mult = Zone(zonePt).Multiplier * Zone(zonePt).ListMultiplier;
+                    PreDefTableEntry(pdchIntDrGrArea, surfName, Surface(iSurf).GrossArea * mult);
+                    PreDefTableEntry(pdchIntDrParent, surfName, Surface(iSurf).BaseSurfName);
+                    computedNetArea(Surface(iSurf).BaseSurf) -= Surface(iSurf).GrossArea * mult;
```
Comment (Nigusse) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:1225]: *Interior Door* data are processed and passed to PreDefTableEntry() function to create the output table here. This looks good.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 3]
```diff
@@ -1238,18 +1276,20 @@ namespace HeatBalanceSurfaceManager {
...
-                isExterior = true;
-                {
-                    auto const SELECT_CASE_var(Surface(iSurf).Class);
-                    if ((SELECT_CASE_var == SurfaceClass_Wall) || (SELECT_CASE_var == SurfaceClass_Floor) || (SELECT_CASE_var == SurfaceClass_Roof)) {
-                        surfName = Surface(iSurf).Name;
-                        PreDefTableEntry(pdchOpNetArea, surfName, computedNetArea(iSurf));
-                    }
+                if ((SurfaceClass == SurfaceClass_Wall) || (SurfaceClass == SurfaceClass_Floor) || (SurfaceClass == SurfaceClass_Roof)) {
+                    surfName = Surface(iSurf).Name;
+                    PreDefTableEntry(pdchOpNetArea, surfName, computedNetArea(iSurf));
```
Comment (Nigusse) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:1260]: Enters the *NetArea* data for interior opaque surfaces. The *NetArea* for interior door goes into a different output table section and is processed separately.

File: src/EnergyPlus/HeatBalanceSurfaceManager.cc [hunk 4]
```diff
@@ -1238,18 +1276,20 @@ namespace HeatBalanceSurfaceManager {
...
                 }
-            }
+            }else {
+                if ((SurfaceClass == SurfaceClass_Wall) || (SurfaceClass == SurfaceClass_Floor) || (SurfaceClass == SurfaceClass_Roof)) {
+                    surfName = Surface(iSurf).Name;
+                    PreDefTableEntry(pdchIntOpNetArea, surfName, computedNetArea(iSurf));
+                }
+            }// interior surfaces
```
Comment (Nigusse) [inline src/EnergyPlus/HeatBalanceSurfaceManager.cc:1267]: The same thing here but *interior door.*

File: src/EnergyPlus/OutputReportPredefined.cc [hunk 5]
```diff
@@ -1704,6 +1736,18 @@ namespace OutputReportPredefined {
...
+        pdchIntOpCons = newPreDefColumn(pdstIntOpaque, "Construction");
+        pdchIntOpRefl = newPreDefColumn(pdstIntOpaque, "Reflectance");
+        pdchIntOpUfactFilm = newPreDefColumn(pdstIntOpaque, "U-Factor with Film [W/m2-K]");
+        pdchIntOpUfactNoFilm = newPreDefColumn(pdstIntOpaque, "U-Factor no Film [W/m2-K]");
+        pdchIntOpGrArea = newPreDefColumn(pdstIntOpaque, "Gross Area [m2]");
+        pdchIntOpNetArea = newPreDefColumn(pdstIntOpaque, "Net Area [m2]");
+        pdchIntOpAzimuth = newPreDefColumn(pdstIntOpaque, "Azimuth [deg]");
+        pdchIntOpTilt = newPreDefColumn(pdstIntOpaque, "Tilt [deg]");
+        pdchIntOpDir = newPreDefColumn(pdstIntOpaque, "Cardinal Direction");
+
```
Comment (Nigusse) [inline src/EnergyPlus/OutputReportPredefined.cc:1750]: *Opaque Interior* surfaces envelope summary output table *title* and *column heading* are set here. All column names and units are consistent with existing envelope summary tables.

File: src/EnergyPlus/OutputReportPredefined.cc [hunk 6]
```diff
@@ -1746,6 +1790,14 @@ namespace OutputReportPredefined {
...
         pdchDrParent = newPreDefColumn(pdstDoor, "Parent Surface");
 
+        pdstIntDoor = newPreDefSubTable(pdrEnvelope, "Interior Door");
+
+        pdchIntDrCons = newPreDefColumn(pdstIntDoor, "Construction");
+        pdchIntDrUfactFilm = newPreDefColumn(pdstIntDoor, "U-Factor with Film [W/m2-K]");
+        pdchIntDrUfactNoFilm = newPreDefColumn(pdstIntDoor, "U-Factor no Film [W/m2-K]");
+        pdchIntDrGrArea = newPreDefColumn(pdstIntDoor, "Gross Area [m2]");
+        pdchIntDrParent = newPreDefColumn(pdstIntDoor, "Parent Surface");
+
```
Comment (Nigusse) [inline src/EnergyPlus/OutputReportPredefined.cc:1800]: *Interior Door* envelope summary output table title and column heading are set here. All column headings and units are consistent with existing tables.

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 7]
```diff
@@ -591,15 +591,19 @@ namespace SurfaceGeometry {
...
+                } else if (SurfaceClass == SurfaceClass_Door) {
+                    PreDefTableEntry(pdchDrUfactFilm, Surface(SurfNum).Name, NominalUwithConvCoeffs, 3);
+                }
+            }else{
+                if ((SurfaceClass == SurfaceClass_Wall) || (SurfaceClass == SurfaceClass_Floor) || (SurfaceClass == SurfaceClass_Roof)) {
+                    PreDefTableEntry(pdchIntOpUfactFilm, Surface(SurfNum).Name, NominalUwithConvCoeffs, 3);
+                } else if (SurfaceClass == SurfaceClass_Door) {
+                    PreDefTableEntry(pdchIntDrUfactFilm, Surface(SurfNum).Name, NominalUwithConvCoeffs, 3);
                 }
             }
```
Comment (Nigusse) [inline src/EnergyPlus/SurfaceGeometry.cc:615]: Surfaces U-values with film are set here but now interior and exterior opaque surfaces are entered into their respective envelope summary table. Similarly the interior and exterior doors U-values with film are entered into their respective envelope summary table.

File: src/EnergyPlus/OutputReportPredefined.hh [hunk 8]
```diff
@@ -198,6 +198,16 @@ namespace OutputReportPredefined {
...
+    extern int pdstIntOpaque;
+    extern int pdchIntOpCons;
+    extern int pdchIntOpRefl;
+    extern int pdchIntOpUfactFilm;
+    extern int pdchIntOpUfactNoFilm;
+    extern int pdchIntOpGrArea;
+    extern int pdchIntOpNetArea;
+    extern int pdchIntOpAzimuth;
+    extern int pdchIntOpTilt;
+    extern int pdchIntOpDir;
```
Comment (Nigusse) [inline src/EnergyPlus/OutputReportPredefined.hh:210]: Looks good. I see consistent variables naming here.

**Comment by Nigusse** [comment]
Code changes look good. I will build the branch locally, check the new unit tests, and will run the unit tests suite serially.

**Comment by Nigusse** [comment]
Interior Opaque Surfaces sample table output. All look good.

IntOpaqueEnvelope

**Comment by Nigusse** [comment]
Interior Door sample table output. All look good.

IntDoor

**Comment by Nigusse** [comment]
Unit tests serial run on local machine passed.

**Comment by Nigusse** [comment]
@Myoldmopar @yzhou601 All code changes and unit test look good except one warning message *"suggest parentheses around ‘&&’ within ‘||’ [-Wparentheses]"* from the CI machine about [this line of code](https://github.com/NREL/EnergyPlus/blob/4eedcdceade236dfcdf3d9b1db7b494a21b904f6/src/EnergyPlus/HeatBalanceSurfaceManager.cc#L1202). But I do not see any problem about the code. See the code spinet below:

`}else if (((Surface(iSurf).Class == SurfaceClass_Window) || (Surface(iSurf).Class == SurfaceClass_TDD_Dome)) && (!dataConstruction.Construct(Surface(iSurf).Construction).TypeIsAirBoundaryInteriorWindow)) {`

Is it complaining about redundant parenthesis for the third conditional test?

**Comment by Myoldmopar** [comment]
It's complaining about the && symbol (the CI warning shows the line number and column number to help narrow that down).  

I will note that the line you showed above does not match the line shown in the link you sent.  You have different parentheses.  Did you edit that line locally but not push it?

Your line:

```
}else if (((Surface(iSurf).Class == SurfaceClass_Window) || (Surface(iSurf).Class == SurfaceClass_TDD_Dome)) && (!dataConstruction.Construct(Surface(iSurf).Construction).TypeIsAirBoundaryInteriorWindow)) {
```

I'm going to stick with the line as it is shown on the link you posted.  The line is:

```
}else if ((Surface(iSurf).Class == SurfaceClass_Window) || (Surface(iSurf).Class == SurfaceClass_TDD_Dome) && (!dataConstruction.Construct(Surface(iSurf).Construction).TypeIsAirBoundaryInteriorWindow)) {
```

But why is it complaining about it?  The answer is the associativity of operator [precedence](https://en.cppreference.com/w/cpp/language/operator_precedence).  That IF condition is comprised of the following segments:

```
(IsWindow) OR (IsTDDDome) AND (NOT AirBoundaryWindow).
```

The warning is because the OR and the AND may be evaluated in an unexpected way at runtime, and it would be better to be very clear by adding parentheses to explicitly specify the order of operations.

**Comment by Nigusse** [comment]
There was a merge conflict at that line that I edited to resolve the conflict. I will push the branch and see how the CI machine responds.

**Comment by Myoldmopar** [comment]
Sounds good, as long as the compiler doesn't see a structure like `if (a AND b OR c)` without any parentheses to clarify the intent, then it should be happy.

**Review by Nigusse — APPROVED** [review]
The code changes and the unit test look good. The unit test demonstrates the feature well. Built the branch code locally and tested interior opaque surface and interior door summary table outputs. All look good. This is ready for merge if the CI machine runs all green.

**Comment by Nigusse** [comment]
Regression run shows that the example file *"1ZoneUncontrolled_FCfactor_Slab_UGWall.idf"* failed. Run the example file locally and confirmed the diffs are caused by re-order of the opaque exterior surface list; otherwise, the values are identical. Thus, the diffs are justified. Note that now underground walls and slabs construction report report U-Factor with film.

ExtOpaqueListReorder

**Comment by Myoldmopar** [comment]
@Nigusse thank you for the review effort.  I am building and running locally with latest develop just to verify, but if all goes well I will merge this.

**Comment by Myoldmopar** [comment]
Everything works locally.  Calling this good.  Thanks @yzhou601 and @Nigusse.

**Pull request merged**